cmd/snap-update-ns/secure_bindmount.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/snapcore/snapd/osutil/sys"
+)
+
+// openNoFollow creates a path file descriptor for the given
+// directory, making sure no components are symbolic links
+func openNoFollow(path string) (int, error) {
+	if !filepath.IsAbs(path) {
+		return -1, fmt.Errorf("path %v is not absolute", path)
+	}
+	segments, err := splitIntoSegments(path)
+	if err != nil {
+		return -1, err
+	}
+	// We use the following flags to open:
+	//  O_PATH: we don't intend to use the fd for IO
+	//  O_NOFOLLOW: don't follow symlinks
+	//  O_DIRECTORY: we expect to find directories
+	flags := sys.O_PATH | syscall.O_NOFOLLOW | syscall.O_DIRECTORY
+	parentFd, err := sysOpen("/", flags, 0)
+	if err != nil {
+		return -1, err
+	}
+	for _, segment := range segments {
+		fd, err := sysOpenat(parentFd, segment, flags, 0)
+		if err != nil {
+			sysClose(parentFd)
+			return -1, err
+		}
+		sysClose(parentFd)
```

Comment:
Nitpick/consistency: we used to have exactly this kind of code but we later on switched to using defer so that we don't have to wonder twice if anything goes wrong. Could you please look at `secureMkPrefix` and put the code just next to it, looking almost the same but not creating the new directories. This will have an impact on tests but will hopefully not change anything major.